lib/versioning/VersionID.ts

```diff
@@ -1,29 +1,67 @@
 // where:
 // - timestamp              14 bytes        epoch in ms (good untill 5138)
 // - sequential_position    06 bytes        position in the ms slot (1B ops)
 // - rep_group_id           07 bytes        replication group identifier
 // - other_information      arbitrary       user input, such as a unique string
+//
+// Legacy Base62 VersionID:
+//         timestamp  sequential_position  rep_group_id
+// where:
+// - timestamp              14 bytes        epoch in ms
+// - sequential_position    06 bytes        position in the ms slot
+// - rep_group_id           07 bytes        replication group identifie
+//
+// Base62 VersionID:
+//         timestamp  sequential_position  rep_group_id  instance_id  version_id_format
+// where:
+// - timestamp              14 bytes        epoch in ms
+// - sequential_position    06 bytes        position in the ms slot
+// - rep_group_id           07 bytes        replication group identifier
+// - instance_id            06 bytes        unique instance identifier
+// - version_id_format      02 bytes        version ID format marker + version
 
 import base62Integer from 'base62';
 import baseX from 'base-x';
+import assert from 'assert';
+import { VersioningConstants } from './constants';
 const BASE62 = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
 const base62String = baseX(BASE62);
 
 // the lengths of the components in bytes
 const LENGTH_TS = 14; // timestamp: epoch in ms
 const LENGTH_SEQ = 6; // position in ms slot
 const LENGTH_RG = 7; // replication group id
+const LENGTH_ID = 6; // instance id
+const LENGTH_FT = 2; // version ID format, 1 byte + separator
 
 // empty string template for the variables in a versionId
 const TEMPLATE_TS = new Array(LENGTH_TS + 1).join('0');
 const TEMPLATE_SEQ = new Array(LENGTH_SEQ + 1).join('0');
 const TEMPLATE_RG = new Array(LENGTH_RG + 1).join(' ');
+const TEMPLATE_ID = new Array(LENGTH_ID + 1).join('0');
+
+export const S3_VERSION_ID_ENCODING_TYPE = process.env.S3_VERSION_ID_ENCODING_TYPE;
+
+// Flag to enable the new version ID (35 characters) over legacy shortID format (27 characters).
+// When enabled and S3_VERSION_ID_ENCODING_TYPE is 'base62':
+//   - Uses new format: timestamp + sequential_position + rep_group_id + instance_id + version_id_format
+//   - Includes instance_id field to differentiate version IDs across multiple instances in the same k8s cluster
+//   - Appends format marker and version identifier for format detection
+// When disabled and S3_VERSION_ID_ENCODING_TYPE is 'base62':
+//   - Uses old format: timestamp + sequential_position + rep_group_id (legacy 27-char format)
+// Falls back to hex encoding if S3_VERSION_ID_ENCODING_TYPE is 'hex' or unset
+export const ENABLE_FORMATTED_VERSION_ID = process.env.ENABLE_FORMATTED_VERSION_ID === 'true';
+
+// version ID format added to the end of the version ID
+const versionIDFormat = '1';
+const vidFormat = `${VersioningConstants.VersionId.FormatMarker}${versionIDFormat}`;
```

Comment:
Use all caps for consistency, and suggest renaming:

```suggestion
const VERSION_ID_FORMAT_VERSION = '1';
const VERSION_ID_FORMAT_SUFFIX = `${VersioningConstants.VersionId.FormatMarker}${VERSION_ID_FORMAT_VERSION}`;
```